File: 프로그래머스/2/131127. 할인 행사/할인 행사.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

int solution(vector<string> want, vector<int> number, vector<string> discount) {
    
    // 1. discount 물건의 map을 선언하고 채운다. (첫날 ~ 열번째 날)
    unordered_map<string, int> discount_map;
    for(string d: discount) discount_map[d] = 0;
    
    string goods;
    for(int i=0; i<10; i++) {
        goods = discount[i];
        discount_map[goods]++;
    }
    
    // 2. i를 0 ~ (discount 날짜 - 10)으로 순회하면서 가능한 날짜를 카운트한다.
    int res = discount.size() - 10;
    int count = 0;
    for(int i=0; i<=res; i++) {
        // i번째 날부터, i+9번째 날의 상품이 가능한지 판별
        // 이 때 원하는 물건과 discount 물건의 map을 비교하면서 판단한다.
        for(int k = 0; k<want.size(); k++) {
            if (discount_map[want[k]] < number[k]) {
                count--;
                break;
            }
        }
        count++;
        if (i == res) break; // 마지막날은 물건 추가 x
        discount_map[discount[i]]--;
        discount_map[discount[i+10]]++;
    }
    
    int answer = count;
    return answer;
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.*;

class Solution {
    
    static Map<String, Integer> dm = new HashMap<>();
    
    public int solution(String[] want, int[] number, String[] discount) {
        
        // 1. 할인하는 물건 목록의 Map을 생성하고, 첫날~10일치를 저장한다.
        String cur;
        for(int i=0; i<9; i++) {
            cur = discount[i];
            if (!dm.containsKey(cur))
                dm.put(cur, 1);
            else
                dm.put(cur, dm.get(cur)+1);
        }
        
        // 2. 날짜가 바뀜에 따라 할인 목록을 갱신하고, 원하는 물건 목록과 비교한다.
        int count = 0;
        for(int i=9; i<discount.length; i++) {
            count++;
            cur = discount[i];
            
            if (!dm.containsKey(cur)) dm.put(cur, 1);
            else dm.put(cur, dm.get(cur)+1);
            
            for(int k=0; k<want.length; k++) {
                if (!dm.containsKey(want[k]) || (dm.get(want[k]) != number[k])) {
                    count--;
                    break;
                }
            }
            
            // 현재로부터 9일 전 상품을 뺀다. (다음 순회에서의 10일 전 상품에 해당)
            dm.put(discount[i-9], dm.get(discount[i-9])-1);
        }

        int answer = count;
        return answer;
    }
}